allow annotation on MachinePool to set externally managed

mweibel opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
As a cluster operator, I can mark a MachinePool with the annotation `cluster.x-k8s.io/replicas-managed-by-autoscaler` to indicate it should not enforce the `MachinePool.replicas` field but instead reconcile it to match the ASGs desiredCapacity field. This allows me to use an external autoscaler which does not come into conflict with the MachinePool reconciliation loop. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2022

**Special notes for your reviewer**:
The annotation is the same as in kubernetes-sigs/cluster-api-provider-azure#2444.

Unfortunately, I did not yet manage to make the added test in `awsmachinepool_controller_test.go` pass. This is also why the PR is marked as WIP/Draft. The test fails with:

```
I0822 10:22:31.013336   20405 awsmachinepool_controller.go:256]  "msg"="Setting MachinePool replicas to ASG DesiredCapacity" "external"=1 "local"=0
    awsmachinepool_controller_test.go:313: 
        Expected
            <errors.aggregate | len:1, cap:1>: [
                <*meta.NoKindMatchError | 0x14000263ec0>{
                    GroupKind: {
                        Group: "cluster.x-k8s.io",
                        Kind: "MachinePool",
                    },
                    SearchedVersions: ["v1beta1"],
                },
            ]
        to be nil
```

I believe the issue is that the API client does not know about the MachinePool CRD. I added a fake machinepool CRD just like the one for Machines/Clusters but it did not resolve the issue. Would be great if somebody could have a look at it. 

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests
- [ ] adds or updates e2e tests

Skarlso commented:
/ok-to-test

sedefsavas commented:
/test ?

sedefsavas commented:
There is a related PR to this in CAPI:
https://github.com/kubernetes-sigs/cluster-api/pull/7107

mweibel commented:
@sedefsavas yes saw that one, thanks. Once that is merged and released, this code could reuse the const from CAPI. 

I still didn't figure out the issue with the failing test. @richardcase had a look at it IIRC but not sure what the state is.

mweibel commented:
/test ?

mweibel commented:
/retest

mweibel commented:
I'm trying every now and then when I find some time to figure out the reason why the MachinePool Kind is not found. I'd love some help on this topic :)

richardcase commented:
I had a quick look at this and i also couldn't work out why it was failing.

Skarlso commented on controllers/suite_test.go:
```diff
@@ -45,6 +47,8 @@ func TestMain(m *testing.M) {
 func setup() {
 	utilruntime.Must(infrav1.AddToScheme(scheme.Scheme))
 	utilruntime.Must(clusterv1.AddToScheme(scheme.Scheme))
+	utilruntime.Must(expclusterv1.AddToScheme(scheme))
```
Shouldn't this be `scheme.Scheme`? Maybe that's what's causing issues.

mweibel replied:
possibly - sorry, my second commit was actually just WIP garbage commit where I tried to find the cause. Will revert those unnecessary changes.

Skarlso commented on controllers/suite_test.go:
```diff
@@ -45,6 +47,8 @@ func TestMain(m *testing.M) {
 	utilruntime.Must(infrav1.AddToScheme(scheme.Scheme))
 	utilruntime.Must(clusterv1.AddToScheme(scheme.Scheme))
+	utilruntime.Must(expclusterv1.AddToScheme(scheme))
+	utilruntime.Must(expinfrav1.AddToScheme(scheme.Scheme))
```
Or this one be also `scheme`. :D I'm actually unsure. :D

Skarlso commented on controllers/suite_test.go:
```diff
@@ -45,6 +47,8 @@ func TestMain(m *testing.M) {
 func setup() {
 	utilruntime.Must(infrav1.AddToScheme(scheme.Scheme))
 	utilruntime.Must(clusterv1.AddToScheme(scheme.Scheme))
+	utilruntime.Must(expclusterv1.AddToScheme(scheme))
```
Yep, it's this. This is actually a compile time failure for me if I check this out.

richardcase replied:
I have it as `scheme.Scheme` and i get the `meta.NoKindMatchError` error. I'm stumped

richardcase commented:
@mweibel - i've got to the bottom of the `NoKindMatchError` error you are seeing. I am going to make a small PR later today. Once that merges you will need to rebase.

richardcase commented:
That change has merged. So if you rebase you should be able to get past the `NoKindMatchError` error.

mweibel commented:
@richardcase oh - that change makes sense :+1: thanks for investigating and fixing it! Will rebase as soon as I'm able (currently on vacation)

Skarlso commented:
/lgtm

Skarlso commented:
/lgtm

richardcase commented:
/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-eks

richardcase commented on pkg/cloud/scope/machinepool.go:
```diff
@@ -43,11 +43,14 @@ import (
 	client.Client
-	patchHelper *patch.Helper
+	patchHelper            *patch.Helper
+	MachinePoolPatchHelper *patch.Helper
```
Ideally we'd want to not export this and provide a `PatchMachinePool` function on the scope. But i'm not so concerned about this

mweibel replied:
I wondered if I should do it that way or provide a specific function on the scope. I don't mind changing this.

mweibel replied:
adjusted :+1:

richardcase commented:
e2e tests are passing :+1:

Skarlso commented:
Ah, that's a good catch. We actually ran into a similar problem with one of the issues. That was patching some template which ended up being out of sync with what you basically had locally. So once you applied again, the changes were lost.

Interesting thought. I don't see an option out of that other than having an unmanaged field perhaps.

mweibel commented:
@richardcase that's a good catch indeed. We're using ArgoCD where you can ignore differences. A user with a GitOps solution would need to do that:
```yaml
  ignoreDifferences:
    - group: cluster.x-k8s.io
      kind: MachinePool
      jsonPointers:
        - /spec/replicas
```

I agree that this is not beautiful. One way this could be different when CAPI would provide e.g. a `status.replicas` field which overrules `spec.replicas` field in case it's set, so `spec.replicas` would only be considered for the first reconciliation. Not sure if that's a good approach though.

dthorsen commented:
An even better option is to completely omit the replicas field from your manifest.

richardcase commented:
> An even better option is to completely omit the replicas field from your manifest.

@dthorsen - i think thats probably the best solution. 

What do you think to the following:

- Change the MachinePool update logic that in the `AWSMachinePool` reconciler so that it only updates the replicas if its not nil
- Add a documentation for externally managed replicas that includes:
  - The option/recommendation of not setting replicas on MachinePool when using externally managed
  - The sample above for ignoring the field with argo

Skarlso commented:
> An even better option is to completely omit the replicas field from your manifest.

Wouldn't that still fail? Because it's nil locally, but set upstream?

richardcase commented:
> Wouldn't that still fail? Because it's nil locally, but set upstream?

Its omit it from the upstream `MachinePool`

AverageMarcus commented on pkg/cloud/scope/machinepool.go:
```diff
@@ -43,11 +43,14 @@ import (
 	"sigs.k8s.io/cluster-api/util/patch"
 )
 
+const ReplicasManagedByAutoscalerAnnotation = "cluster.x-k8s.io/replicas-managed-by-autoscaler"
```
Am I correct that we'd import this value from CAPI once [this PR](https://github.com/kubernetes-sigs/cluster-api/pull/7107) is merged rather than hardcoding it in here?

mweibel replied:
yes, I think that would make sense.

mweibel commented:
> > Wouldn't that still fail? Because it's nil locally, but set upstream?
> 
> Its omit it from the upstream `MachinePool`

@richardcase looking at the webhook code, I believe this wouldn't work since the webhook would set it anyway:
https://github.com/kubernetes-sigs/cluster-api/blob/main/exp/api/v1beta1/machinepool_webhook.go#L54..L56

or do I miss something there?

mweibel commented:
I adjusted the patch helper so it's a function on the MachinePool scope instead of exposing the patch helper itself. 

@richardcase how to handle MachinePool.replicas: I don't believe there is currently a better option. I added some documentation which might help in this case. 

Would be great to have another review on this :+1:

AverageMarcus commented on docs/book/src/topics/machinepools.md:
```diff
@@ -104,4 +104,47 @@ spec:
+## Autoscaling
+
+[`cluster-autoscaler`](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) can be used to scale MachinePools up and down.
+Two providers are possible to use with CAPZ MachinePools: `clusterapi`, or `aws`.
```
```suggestion
Two providers are possible to use with CAPA MachinePools: `clusterapi`, or `aws`.
```

Ankitasw replied:
Good catch 😄

mweibel replied:
oh 😄 sorry about that one! good catch indeed.

mweibel commented:
docs typo is fixed 👍

mweibel commented:
FTR this PR has been updated with the updated annotation key/value from kubernetes-sigs/cluster-api#7107.

mweibel commented:
/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-eks

AverageMarcus commented:
/lgtm

Skarlso commented:
/approve
